### ISSUE TITLE
MacBookPro10,1: No suitable _DSM call found. (→Use gmux  instead of bbswitch)

### ISSUE THREAD

**Issue by bukzor** [Bumblebee-Project/Bumblebee#585]
I believe the most relevant error message is this one:

``` messages
[  457.644525] bbswitch: version 0.8
[  457.644538] bbswitch: Found integrated VGA device 0000:00:02.0: \_SB_.PCI0.IGPU
[  457.644547] bbswitch: Found discrete VGA device 0000:01:00.0: \_SB_.PCI0.P0P2.GFX0
[  457.644563] ACPI Warning: \_SB_.PCI0.P0P2.GFX0._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20131115/nsarguments-95)
[  457.644612] ACPI Warning: \_SB_.PCI0.P0P2.GFX0._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20131115/nsarguments-95)
[  457.644637] ACPI Warning: \_SB_.PCI0.IGPU._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20131115/nsarguments-95)
[  457.644657] bbswitch: No suitable _DSM call found.
```

I've pasted the results of bumblebee-bugreport here:
https://gist.github.com/bukzor/38e2152a4b670661f37d

```
DISTRO:
Distributor ID: Ubuntu
Description:    Ubuntu 14.04 LTS
Release:    14.04
Codename:   trusty

KERNEL:
Linux xubuntu-try4 3.13.0-29-generic #53-Ubuntu SMP Wed Jun 4 21:00:20 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

DMIDECODE:
baseboard-manufacturer: Apple Inc.
baseboard-product-name: Mac-C3EC7CD22292981F
baseboard-version     : MacBookPro10,1
system-manufacturer   : Apple Inc.
system-product-name   : MacBookPro10,1
system-version        : 1.0
bios-vendor           : Apple Inc.
bios-version          : MBP101.88Z.00EE.B02.1208081132
bios-release-date     : 08/08/2012

VERSIONS:
bumblebeed (Bumblebee) 3.2.1
```

**Comment by benklop** [comment]
I've got a macbook mid-2014 and I'm not sure it's actually still possible to use the gmux in place of bbswitch, as of Ubuntu 16.04 with a 4.4.0 kernel. The vga-switcheroo interface in /sys/kernel/debug seems to be missing, and I've got all the necessary modules loaded.

**Comment by Lekensteyn** [comment]
bbswitch does not support macbooks. Paging in the expert @l1k

**Comment by l1k** [comment]
The `switch` file in debugfs will appear as soon as the modules nouveau, i915 and apple-gmux are loaded. If it's missing, you haven't loaded the drivers or they failed for some reason (look in dmesg). On the MacBookPro11,3 (Mid 2014) you need either a kernel patch or bootloader support to identify as Apple MacOS X 10.9 to EFI, otherwise the integrated GPU is turned off by EFI when ExitBootServices is called (BootCamp mode for Windows). See https://github.com/0xbb/gpu-switch#macbook-pro-113-and-115-notes .

On retinas (including the MacBookPro10,1 and MacBookPro11,3 (Mid 2014) mentioned above), you can turn the discrete GPU on/off but you can't switch (screen is black) because the inactive GPU wasn't able to link-train the eDP output. I've been hacking on some prototypes to make it work but haven't gotten around to finalize it yet. I don't have such a machine myself.

On pre-retinas (including the MacBookPro9,1 mentioned in BumbleBee-Project/bbswitch#125), switching is supported since 4.6.

**Comment by benklop** [comment]
@l1k thanks for the info - I suppose that means that using the Nvidia binary driver isn't supported with bumblebee on an MBP? That's probably the issue I'm facing, as I was trying to not use nouveau. 

I'm using a hack [gpu-switch](https://github.com/0xbb/gpu-switch) to set the correct GPU then reboot when necessary. If you need help testing your prototypes, I'd be happy to offer assistance.

**Comment by l1k** [comment]
@benklop: Yes, the nvidia driver isn't registering with vga_switcheroo then. Unless they add that, all bets are off.

gpu-switch only influences the switch state as set by the gmux EFI driver, but still both GPUs will be turned on. To turn the discrete GPU off and reduce power consumption, you can echo OFF to the vga_switcheroo debugfs interface but again, for it to appear nouveau is required...
